crates/rome_formatter/src/lib.rs

```diff
@@ -152,26 +154,254 @@ impl Default for FormatOptions {
     }
 }
 
+/// Lightweight sourcemap marker between source and output tokens
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SourceMarker {
+    /// Position of the marker in the original source
+    pub source: TextSize,
+    /// Position of the marker in the output code
+    pub dest: TextSize,
+}
+
 #[derive(Debug, Clone, Eq, PartialEq)]
 pub struct Formatted {
     code: String,
+    range: Option<TextRange>,
+    sourcemap: Vec<SourceMarker>,
 }
 
 impl Formatted {
-    pub fn new(code: String) -> Self {
-        Self { code }
+    fn new(code: String, range: Option<TextRange>, sourcemap: Vec<SourceMarker>) -> Self {
+        Self {
+            code,
+            range,
+            sourcemap,
+        }
+    }
+
+    fn empty() -> Self {
+        Self {
+            code: String::new(),
+            range: None,
+            sourcemap: Vec::new(),
+        }
+    }
+
+    /// Range of the input source file covered by this formatted code,
+    /// or None if the entire file is covered in this instance
+    pub fn range(&self) -> Option<TextRange> {
+        self.range
+    }
+
+    pub fn sourcemap(&self) -> &[SourceMarker] {
+        &self.sourcemap
     }
 
     pub fn code(&self) -> &String {
         &self.code
     }
+
+    pub fn into_code(self) -> String {
+        self.code
+    }
 }
 
 /// Formats a JavaScript (and its super languages) file based on its features.
 ///
 /// It returns a [Formatted] result, which the user can use to override a file.
 pub fn format(options: FormatOptions, syntax: &SyntaxNode) -> FormatResult<Formatted> {
-    Formatter::new(options).format_root(syntax)
+    let element = Formatter::new(options).format_root(syntax)?;
+    Ok(Printer::new(options).print(&element))
+}
+
+/// Formats a range withing a JavaScript file
+///
+/// It returns a [Formatted] result with a range corresponding to the
+/// range of the input that was effectively overwritten by the formatter
+pub fn format_range(
+    options: FormatOptions,
+    root: &SyntaxNode,
+    range: TextRange,
+) -> FormatResult<Formatted> {
+    // Find the tokens corresponding to the start and end of the range
+    let mut start_token = None;
+    let mut end_token = None;
+
+    for token in root.descendants_tokens() {
+        let range = token.text_range();
+
+        if start_token.is_none() && range.contains_inclusive(range.start()) {
+            start_token = Some(token.clone());
+        }
+
+        if end_token.is_none() && range.contains_inclusive(range.end()) {
+            end_token = Some(token);
+        }
+
+        if start_token.is_some() && end_token.is_some() {
+            break;
+        }
+    }
+
+    // If not token were found, this means the input node was empty
+    // or the formatting function was not passed the root SyntaxNode
+    // for the file: default to the first and last token in the root
+    // node or exit early with an empty result if no such token exists
```

Comment:
```suggestion
    // If no tokens were found, this means the input node was empty
    // or the formatting function was not passed the root [SyntaxNode]
    // of the file: default to the first and last tokens in the root
    // node, or exit early with an empty result if these tokens don't exist
```

This `or the formatting function was not passed the root [SyntaxNode] of the file` is difficult to understand. Can we reword it?